### PR TITLE
Mirror FTP directories to a GCS path in parallel

### PR DESCRIPTION
**Pull request by kvg**
This workflow mirrors FTP directories to a specified GCS path, optionally placing the files in a directory based on the basename of the FTP directories.  The parallelism level of the mirroring operation is controllable with the 'num_simultaneous_downloads' setting so as not to accidentally DDOS the server.

A test for the workflow is in place (see `test/test_json/DownloadFromFTP.test.json`), which downloads some small files from the 1000 Genomes FTP site.

### CODE REVIEW

**Review by jonn-smith — CHANGES_REQUESTED** [review]
A few questions and comments

File: docker/lr-cloud-downloader/environment.yml [hunk 1]
```diff
@@ -4,7 +4,8 @@ channels:
   - conda-forge
   - bioconda
 dependencies:
-  - sra-tools=2.9.1
-  - entrez-direct=7.70
-  - parallel-fastq-dump=0.6.3
-  - readline=6.2
+  - sra-tools
```
[the remark this answers is not in the document]
Comment (kvg) [reply docker/lr-cloud-downloader/environment.yml]: I've used the 'conda env export' functionality to make the environment very specific and stable.  We should adopt this as best practices for all of our Docker images.

File: wdl/DownloadFromFTP.wdl [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    input {
+        Array[String] ftp_dirs
+
+        RuntimeAttr? runtime_attr_override
+    }
+
+    command <<<
+        set -euxo pipefail
+
+        for FTP_DIR in ~{sep=' ' ftp_dirs}
```
[the remark this answers is not in the document]
Comment (kvg) [reply wdl/DownloadFromFTP.wdl]: I've changed this now to read one line at a time from a temp file.

File: wdl/DownloadFromFTP.wdl [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        do
+            lftp -e 'find -l >> all_files.txt; exit' $FTP_DIR
```
[the remark this answers is not in the document]
Comment (kvg) [reply wdl/DownloadFromFTP.wdl]: I think for now, we can forego it; I haven't been downloading anything from servers that require authentication.  But in the future when this problem comes up, we'll of course have to make the necessary modifications.

File: wdl/DownloadFromFTP.wdl [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+        awk 'NR % ~{num_jobs} == ~{nth} { print $0 }' ~{manifest} | while read line; do
+            ftp_dir=$(echo $line | awk '{ print $1 }')
+            ftp_base=$(basename $ftp_dir)
+            ftp_file=$(echo $line | awk '{ print $2 }' | sed 's/^\.\///')
+
+            gcsfile="~{gcs_out_root_dir}/~{true="$ftp_base/" false="" prepend_dir_name}$ftp_file"
+
+            if gsutil ls $gcsfile ; then
+                echo "$gcsfile already exists."
```
[the remark this answers is not in the document]
Comment (kvg) [reply wdl/DownloadFromFTP.wdl:198]: Provided the files were uploaded by this tool, `gsutil cp` will automatically compute and compare md5 sums.

However, we do technically have an issue where the download from the FTP site might be corrupted, in which case, the upload to GCS will also be broken.  That's a problem I think worth thinking about in a future improvement to this tool.

File: wdl/DownloadFromFTP.wdl [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+            else
+                mkdir -p $(dirname $ftp_file)
+                lftp -e "get $ftp_file -o $ftp_file; exit" $ftp_dir
+
+                gsutil -m cp $ftp_file $gcsfile
+                rm $ftp_file
+
+                RET=1
```
[the remark this answers is not in the document]
Comment (kvg) [reply wdl/DownloadFromFTP.wdl]: This was on purpose but admittedly lazy.  The logic was to let jobs that needed to download files automatically fail, and have the automatically retried job verify the files had been correctly downloaded and placed in GCS.  But that requires using up a retry for no good reason.  I've changed it to now check the return code of `lftp` and `gsutil cp`, and only set RET=1 if one of those tasks fails.  This should eliminate the need for a job to deliberately fail just so the uploads to GCS could be verified.

File: wdl/DownloadFromFTP.wdl [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+        num_simultaneous_downloads: "The number of files to fetch simultaneously."
+        prepend_dir_name:           "If true, place the files in a subdirectory based on the basename of the FTP dir."
+        gcs_output_dir:             "GCS path for storing output"
+    }
+
+    String outdir = sub(gcs_out_root_dir, "/$", "")
+
+    call GetFileManifest { input: ftp_dirs = ftp_dirs }
+
+    scatter (n in range(num_simultaneous_downloads)) {
```
[the remark this answers is not in the document]
Comment (kvg) [reply wdl/DownloadFromFTP.wdl:36]: Added comments now to clarify how this tool works.

**Review by jonn-smith — APPROVED** [review]
One quick comment about adding an issue into the repo.  The rest looks good.

**Pull request merged**